### ISSUE TITLE
Unified StockIcons solution

### ISSUE THREAD

**Issue by sevoku** [mono/xwt#470]
Since we are still missing many stock icons and the declared ones are not available on all backends we should think about a unified solution for this.

Current state: Wpf and Mac have only the small system sets. Gtk has much more, but the stock-item things have been deprecated (and will be removed some day).

One solution could be to remove stock icons completely, since we'll never find more matching system icons on Wpf/Mac (or they change their design concept).

An other way could be to add backend independent icons, what I personally would prefer. We could include the full [Adwaita](https://github.com/GNOME/adwaita-icon-theme) icon set (the icons already included, like list-add.png in Wpf, seem to come from Adwaita). And Adwaita has a compatible [CCBYSA3](http://creativecommons.org/licenses/by-sa/3.0/) license (a reference should be added anyway, if we keep the current state!)

Additionally StockIcon override methods could be implemented. Something like ```Application.SetStockIcon(string id, Image icon)``` and ```Application.SetStockIconFallback(string id, Image icon)``` which can be used to set a fallback icon (used in case the backend has no matching icon) or to completely override the backend.

**Comment by slluis** [comment]
I think we should remove the stock icons, or reduce them to a very limited set. The original idea was to provide icons for common commands and for message dialogs, but I don't think that's very useful anymore. I also don't think it is a good idea to use a platform-independent icon set, because it will look weird in all platforms.

**Comment by sevoku** [comment]
I agree, one default set for all, would look strange. But there are some common icons that are very often used and its more simple to load a stock icon (current API) then to load one from a resource (unique enum id vs. string). And I wouldn't drop the original idea, since it will be much more useful for a Toolbar backend (some day).
The current set is already (or still?) very limited. And I think we can't simply remove StockIcons, since at least the MessageDialogs would not work without them (we could make them internal, though). So why not extend it with more common icons (like file, folder, etc) and add the proposed functions (an ```Application.SetStockIcons(IEnumerable<StockIconId id, Image icon> inconSet)``` would be nice too, to override many icons at once, and we could add a function for testing if a toolkit provides a special stock icon)? To avoid strange looking apps, we could really remove wrong icons (wpf has some) and throw the NotImplemented exception (like before, maybe with a better description) without adding one default set.

It could be a round thing, if it we add good documentation about missing icons and how to add them. And the existing ones make an app really more native looking.

To summerize the idea:
* Add more useful stock ids
* remove all custom icons from code (Wpf/Mac) and leave those loaded from the system
* new functions to handle stock icons (static in ```Application``` or ```Toolkit``` members):
```c#
bool HasStockIcon (StockIconId id)
SetStockIcon (StockIconId id, Image icon, bool overrideToolkit = false)
SetStockIcons(IEnumerable<StockIconId id, Image icon> inconSet, bool overrideToolkit = false)
ResetStockIcon (StockIconId id)
ResetStockIcons ()
```
If the user wants the most native look, he/she can set an own default icon set without overriding the icons from the toolkit. Or simply replace all existing sock icons for a custom corporate design on all systems. Or one can make it depend on the loaded ```ToolkitType```. So more design flexibility without loosing the feature completely.

**Comment by lytico** [comment]
maybe this could be usefull:
https://github.com/mono/xwt/pull/477

**Comment by slluis** [comment]
The only advantage I see about stock icons is that it allows you to write apps whose icons will (or should) look native in each platform. For example, if you want to show a tree view with folders, you could use a stock folder icon that would look like a explorer folder in Windows, or a Finder folder on Mac. So I don't really see the benefit of being able to override the default stock icons. The stock icons are just Image references. If you want to use some other image, just reference your own images, not the stock ones.

If we keep using stock icons, we should define only those for which we can provide a native look in each platform. Ideally we should use stock icons from the system, but some platforms (such as Windows for example) will have a very limited set. We can then provide custom icons, but the danger is then that they can look obsolete when running in newer versions of the OS.

In don't think stock icons will be very useful beyond simple uses like the above. The problem is that the use of stock icons is after all very platform dependent. A toolbar in Windows looks completely different than a toolbar on Mac. Windows shows icons in menus, while Mac doesn't. Even in a platform, the look of an icon may be different depending on the context. On Mac, icons shown in a toolbar have a very specific style, so the same "stock icon" may look different depending on where you show it.

**Comment by sevoku** [comment]
So, if we provide only a minimal set of only available icons on all systems without any additional api, the user has to simply use custom icons. If the user wants to be the icons native looking, he must check the environment and load special icon sets for the different platforms. To do so, there are two possible ways, adding icon resources/loading from filesystem, or using system specific APIs to load the icons from the system resources.
The first way is pretty simple but the user has to take care of theming etc. The second way needs system specific library references (to load a gtk theme icon, the user must reference at least Xwt.Gtk and gtk-sharp, on Windows he must use the Windows API, ...) and good toolkit knwoledge.
So from the users point of view, it would be a big help, if Xwt would have some api to do such things for me, so I have not to mess around with different icons/images for different toolkits (at least not for special icons). And for different purposes, there could be special ids like StockIcons.{name}Symbolic or StockIcons.{name}[Menu|Button|...].

An other way could be to provide some generic ```Image.FromToolkit(params [] object)``` which would load an icon with some toolkit specific parameters like icon name on Gtk, file path and resource id on windows, or a framework name and the icon id on Mac, so the user must still take care of platform detection, but must not load platform specific assemblys (or use the platform APIs). Then we could remove stock icons from the public api. This would make the implementation on Xwt much more simple, but such a function must be well documented.
